truffleruby/src/main/java/org/truffleruby/core/binding/BindingNodes.java

```diff
@@ -99,9 +76,31 @@ public static FrameDescriptor getFrameDescriptor(DynamicObject binding) {
         return Layouts.BINDING.getFrame(binding).getFrameDescriptor();
     }
 
+    public static FrameDescriptor getExtrasDescriptorOrNull(DynamicObject binding) {
+        assert RubyGuards.isRubyBinding(binding);
+        final MaterializedFrame frame = Layouts.BINDING.getExtras(binding);
+        return frame == null ? null : frame.getFrameDescriptor();
+    }
+
+    public static FrameDescriptor getExtrasDescriptor(RubyContext context, DynamicObject binding) {
+        assert RubyGuards.isRubyBinding(binding);
+        MaterializedFrame frame = getExtrasFrame(context, binding);
+        return frame.getFrameDescriptor();
+    }
+
     public static MaterializedFrame getDeclarationFrame(DynamicObject binding) {
         assert RubyGuards.isRubyBinding(binding);
-        return RubyArguments.getDeclarationFrame(Layouts.BINDING.getFrame(binding));
+        return Layouts.BINDING.getFrame(binding);
+    }
+
+    public static MaterializedFrame getExtrasFrame(RubyContext context, DynamicObject binding) {
+        assert RubyGuards.isRubyBinding(binding);
+        MaterializedFrame frame = Layouts.BINDING.getExtras(binding);
+        if (frame == null) {
+            frame = Truffle.getRuntime().createMaterializedFrame(new Object[0], newFrameDescriptor(context));
```

Comment:
Creating a new FrameDescriptor for each Binding means you can only specialize on a single given Binding instance, not multiple having the same set of Binding variables.
OTOH it is safe and fixes the problem mentioned in the previous PR about sharing descriptors.
This is more than fine for now, but for the record that's why DynamicObject could be nice here to be able to specialize on accessing in-binding variables on the fast path from different bindings having the same set of variables, but that's more complex due to keeping a Shape => FrameDescriptor cache and converting DynamicObject to frames on the fast path.